enhancements/network/nodeport-range-configuration.md

```diff
@@ -0,0 +1,125 @@
+a slice of port ranges. It defaults currently to a single port range (30000-32767).
+It is consumed in Openshift via the ConfigMap resource that holds all the api-server configuration.
+In order to allow customers to scale the number of node-port services in the cluster to a value greater
+than that allowed by the default service node-port range, it is important to allow dynamically changing this
+range to allow for more ports. This enhancement focuses on the mechanics of allowing users to configure a
+custom service node-port range for their clusters.
+
+### Goals
+
+- Customers can configure the service node-port range for their cluster during install.
+- Customers can configure the service node-port range for their cluster post install as their needs change.
+
+### Non-Goals
+
+- Services launched prior to customers changing the service node-port range will not work if the new range is
+  not inclusive of the old range.
+- This enhancement makes no guarantees of providing non-disruptive change to the service node-port range.
+  In other words customers, when they make non-inclusive changes to the service node-port range are responsible
+  for deleting old services and recreating new services that will then get node ports assigned from the
+  new range.
+- Configuration of multiple service node-port ranges even though supported by the Kubernetes API server will not be
+  allowed. Users are responsible for configuring the range to be large enough for their needs.
+  
+## Proposal
+
+The config observation facility in cluster-kube-apiserver-operator will be used to observe the openshift/api's network 
+configuration object for changes to the service node port range. The config observation controller will then apply the
+change to the `apiServerArguments.service-node-port-range` field. There will only be one entry allowed in the slice.
+
+The following changes are proposed and implemented:
+
+- API changes: [PR #660](https://github.com/openshift/api/pull/660)
+- Library-go changes: [PR #826](https://github.com/openshift/library-go/pull/826)
+- cluster-kube-apiserver-operator changes: [PR #894](https://github.com/openshift/cluster-kube-apiserver-operator/pull/894)
+
+Additionally, to realize the configuration of service node port range, the necessary ports need to be opened on the 
+individual nodes. Traditionally, this has been the responsibility of the Openshift Installer. But since we want this
+configuration to be changed post-install, we need some mechanism to allow for a new set of ports to be opened by _an_
+_entity_ that knows how to tweak the infrastructure.
+
+### Implementation Details/Notes/Constraints
+
+#### Early CI
+
+#### Upstream Kubernetes
+
+This feature is already supported in upstream Kubernetes.
+
+#### Installer
+
+The installer has been traditionally responsible to [open the ports](https://github.com/openshift/installer/blob/master/data/data/aws/vpc/sg-worker.tf#L146)
```

Comment:
I don't see any discussion of UPI vs IPI and existing clusters - what happens when a UPI cluster from an older release is upgraded and someone tries to set this?

Reply:
Not sure what we currently do for UPI when users make any of the allowed changes post install that need underlying infrastructure changes. If this is a matter of simply documenting that the users are on their own responsible to open additional ports on their infrastructure, that can be added easily to the docs. For IPI, my initial thought was to feed the machine-api operator with a CR that it will reconcile to open the required ports.

Reply:
I am unsure how the machine-api distinguishes between a UPI install and an IPI install.

Reply:
FYI, for 4.6, the customers who need this are using UPI, so this is a little less urgent but we definitely need to look at the security groups issue as a whole since this has come up in other contexts as well for CNO to manage.

Reply:
For 4.6, can we simply support UPI and refuse to allow the range to be overridden (and set a status).  I suspect that the only people who will need to be able to expose more services than the default are UPI setups.

Reply:
How do we refuse the range to be over-ridden in case of IPI? Do we know how to make the distinction? Are you saying only allow it if the platform is UPI?

Reply:
Yes, I was proposing we detect if we are running on an IPI and, since the installer is setting up the exposed ports, and that is hard to change at the moment, then we should just alert and ignore the user setting.